core/src/main/scala/org/apache/spark/scheduler/TaskResultGetter.scala

```diff
@@ -102,6 +102,10 @@ private[spark] class TaskResultGetter(sparkEnv: SparkEnv, scheduler: TaskSchedul
               (deserializedResult, size)
           }
 
+          // quickly return if the task has finished
+          if (scheduler.isFinishedTask(taskSetManager, tid)) {
+            return
```

Comment:
This's only useful when you hit the race condition, which is only a corner case. And in SPARK-37300, I think our target is to fix the bug issue but not for improvement. I personally think this introduces more complexity compared to the benefit we could get.

Reply:
Agree with @Ngone51, this is something which is checked below (in `handleSuccessfulTask`) - the potential improvement, for an unlikely corner case, is not worth the additional complexity.